RHDEVDOCS-3591: odo 250 installation guide

gabriel-rh opened the pull request:
4.6+

[RHDEVDOCS-3591](https://issues.redhat.com/browse/RHDEVDOCS-3591)

[Netlify preview](https://deploy-preview-40570--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/developer_cli_odo/installing-odo.html)

@Preeticp @anandrkskd @kadel

dharmit reviewed (commented):
@gabriel-rh, this looks good to me. But I have a few questions (not requesting changes).
* Since it's d.o.c, should installation instructions for RHEL be at the top instead of bottom?
    * If this makes sense, should "Installing odo on Linux" be reworded to something like "Installing odo on other Linux distributions"?
* The table for "Installing odo on Linux" section looks neat. Why not have that for Windows & macOS as well? My reasoning for this is that, as a user, searching for the correct binary on the [content gateway](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/) is not a very pleasant experience.
* Just like you have used the link for "OpenShift VS Code extension" in "Installing odo on VS Code" section, could you use that for the [content gateway link](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/) as well? Unless, of course, there's a reason to use the full URL

kadel commented:
> Since it's d.o.c, should installation instructions for RHEL be at the top instead of bottom?

I think that the order is fine. I would keep it as it is. Almost no-one uses odo on RHEL why have it on a first place?
The order that Gabriel has there actually matches the usage that we see in our telemetry.

kadel commented:
/lgtm

anandrkskd commented:
/lgtm

dharmit commented:
> Almost no-one uses odo on RHEL why have it on a first place?

My reason to bring that up was that this PR is for docs.openshift.com and not for odo.dev.

kadel commented:
> My reason to bring that up was that this PR is for docs.openshift.com and not for odo.dev.

Even OpenShift customers are not going to run RHEL on developer laptops.
Even documentation for `oc` doesn't have instructions for RHEL, they only generic Linux instructions.

jc-berger commented on cli_reference/developer_cli_odo/installing-odo.adoc:
```diff
@@ -6,23 +6,23 @@ include::modules/common-attributes.adoc[]
 toc::[]
 
-The following section describes how to install `{odo-title}` on different platforms using the CLI or the Visual Studio Code (VS Code) IDE.
+// The following section describes how to install `{odo-title}` on different platforms using the CLI or the Visual Studio Code (VS Code) IDE.
```
IBMG SG pg 22: avoid self referential language, like `This section describes` 

Could you simply say, `You can install {odo-title} on different platforms using the CLI or the Visual Studio Code (VS Code) integrated development environment (IDE)` or `Install {odo-title} on different platforms...`

Note, if this is a standard intro that openshift  docs use, don't change it. You want to be consistent with the other docs' sections. But if you do have the leeway to change it, consider dropping the self referential language.

jc-berger replied:
Also, for some reason, I don't see this intro sentence in the preview link:
<img width="1331" alt="Screen Shot 2022-01-19 at 2 36 35 PM" src="https://user-images.githubusercontent.com/70717303/150201518-2e69ae2a-d44e-49ac-bc57-f353e0c3df7e.png">

gabriel-rh replied:
The line including "The following section describes..." is actually commented out in the source for this precise reason - I will delete the comment entirely, for clarity

jc-berger commented on cli_reference/developer_cli_odo/installing-odo.adoc:
```diff
@@ -6,23 +6,23 @@ include::modules/common-attributes.adoc[]
-The following section describes how to install `{odo-title}` on different platforms using the CLI or the Visual Studio Code (VS Code) IDE.
+// The following section describes how to install `{odo-title}` on different platforms using the CLI or the Visual Studio Code (VS Code) IDE.
+
+You can install the `{odo-title}` CLI on Linux, Windows, or macOS by downloading a binary. You can also install the OpenShift VS Code extension, which uses both the `{odo-title}` and the `oc` binaries to interact with your OpenShift Container Platform cluster. For {op-system-base-full}, you can install the `{odo-title}` CLI as an RPM.
```
You tell us the benefit of installing the OpenShift VS Code extension: it uses both the odo and oc binaries and interacts with the user's OpenShift Container Platform cluster. 

So, what's the benefit of installing the odo CLI as an RPM? If you give users multiple options, give them a benefit for each option (like you did for the OpenShift VS Code extension). Basically, right now, I don't know why I would want/consider installing the odo cli as an RPM. 

Also, tell us what RPM is, then put the acronym in parenthesis since this is our first time seeing the acronym.

jc-berger commented on cli_reference/developer_cli_odo/installing-odo.adoc:
```diff
@@ -6,23 +6,23 @@ include::modules/common-attributes.adoc[]
 ====
 
-You can also find the URL to the latest binaries from the {product-title} web console by clicking the *?* icon in the upper-right corner and selecting *Command Line Tools*
+// You can also find the URL to the latest binaries from the {product-title} web console by clicking the *?* icon in the upper-right corner and selecting *Command Line Tools*
```
Once again, I don't see this in the preview:
<img width="1332" alt="Screen Shot 2022-01-19 at 2 41 14 PM" src="https://user-images.githubusercontent.com/70717303/150202124-a418adcb-2f6f-4862-822b-0d92af1d97f4.png">


The sentence is good, but for accessibility, we want to avoid words with directions, like top, bottom, left, right. So consider a simple rewrite:

`...web console by clicking the *7* icon and selecting *Command Line Tools*.`

gabriel-rh replied:
I have explicitly commented out this option, because it currently leads you to an older version of odo, namely 2.4.3 and to a different mirror, rather than the content gateway. Once with is updated in a newer release of OpenShift, I will restore the instructions.

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux-rpm.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.Procedure
+
+. Register with Red Hat Subscription Manager:
```
Consider giving us more of a reason as to why we'd want to install the odo cli as RPM. Something simple like this formula: 
`To <state benefit/why user would care> on Red Hat Enterprise Linux (RHEL), install the odo CLI as a RPM.` 

I'm not sure of the reason, but maybe the benefit is something like this:
`To work with simple binaries in complex operating systems on RHEL, install the odo CLI as a RPM.`

gabriel-rh replied:
As before, I am following the conventions in oc docs, and they do not explain about RPMs, as it is an assumed knowledge.

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux-rpm.adoc:
```diff
@@ -0,0 +1,56 @@
+# subscription-manager register
+----
+
+. Pull the latest subscription data:
```
Great procedure steps! Succinct, good action verbs, I love them!

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux-rpm.adoc:
```diff
@@ -0,0 +1,56 @@
+# yum install odo
+----
+
+. Verify that `{odo-title}` is now available on your system:
```
I like the inclusion of the verification step. Verification steps are often not included but If ind them helpful, good idea including it!

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux.adoc:
```diff
@@ -6,88 +6,40 @@
 .Procedure
 
-. Obtain the tarball:
+. Navigate to link:https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/[] and download the appropriate file for your operating system and architecture.
```
Give the link a name. Something like, 

Navigate to the page, `link:<link>[odo latest]`, and download the appropriate file for your operating system and architecture.

gabriel-rh replied:
Agreed. I have reworked the section to add table to downloads, clarify binary vs tarball differences

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux.adoc:
```diff
@@ -6,88 +6,40 @@
 
-. Obtain the tarball:
+. Navigate to link:https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/[] and download the appropriate file for your operating system and architecture.
+. If you downloaded a tarball, extract the binary:
```
So, are steps 2-5 all reliant on the user downloading a tarball instead of a different odo file?

If so, then this whole procedure should be focused on users downloading a tarball. In which case, it would go something like this:

1. Navigate to the page `link:<link>[odo latest], and download the file, `tarball.` 
2. Extract the binary:
3. Change the permissions on the binary: 
So on and so forth. 

If users can download whatever file they want, and only step 2 is contingent on them installing a tarball, then step 2 is optional (because they only do this one step if it they downloaded a tarball). So that would look something link:

1. Navigate to the page, `link:<link>[odo latest]`, and download the appropriate file for your operating system and architecture.

- If you downloaded a tarball, extract the binary:

2. Change the permissions on the binary. 
So on and so forth. 

Honestly, reading the procedure, I think the whole thing is expecting users to download the tarball file, so I would simply instruct them to do so in step one.

gabriel-rh replied:
Agreed. I have reworked the section to add table to downloads, clarify binary vs tarball differences

Agreed about only using the tarball option, but I inherited these docs and that is the pattern used. I will propose removing the binary download documentation in a subsequent JIRA, but there may be good reasons for keeping them...

jc-berger commented on modules/developer-cli-odo-installing-odo-on-linux.adoc:
```diff
@@ -6,88 +6,40 @@
-.Procedure
-
-. Register with Red Hat Subscription Manager:
+. Place the `{odo-title}` binary in a directory that is on your `PATH` and rename it to `odo`, if necessary.
```
I would eliminate the `if necessary` users can make that choice. 

`Place the odo binary in a directory that is on your PATH and rename it to odo.`

gabriel-rh replied:
I have reworked this so that the renaming is done earlier in the procedure.

I have copied the pattern for checking the PATH from the oc docs, so they are consistent with those.

jc-berger commented on modules/developer-cli-odo-installing-odo-on-macos.adoc:
```diff
@@ -6,38 +6,34 @@
 = Installing {odo-title} on macOS
 
-== Binary installation
+The `{odo-title}` CLI for macOS is available to download as a binary and as a tarball.
```
See all my comments on `Installing odo on Linux` and apply them here to `Installing odo on MacOS`

gabriel-rh replied:
Agreed - same updates made across all OSes

jc-berger commented on modules/developer-cli-odo-installing-odo-on-windows.adoc:
```diff
@@ -6,29 +6,23 @@
-. Select *Edit environment variables for your account*.
-. Select *Path* from the *Variable* section and click *Edit*.
-. Click *New* and type `C:\go-bin` into the field or click *Browse* and select the directory, and click *OK*.
+The `{odo-title}` CLI for windows is available to download as a binary and as an archive.
```
See all my comments on `Installing odo on Linux` and `Installing odo on MacOS` and apply them here to `Installing odo on Windows`.

gabriel-rh replied:
I have made the changes across all OSes - so they should now be consistent and updated appropriately

jc-berger commented on modules/developer-cli-odo-installing-odo-on-windows.adoc:
```diff
@@ -6,29 +6,23 @@
+.Procedure
+
+. Navigate to link:https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/[] and download the `odo-windows-amd64.exe` binary or the `odo-windows-amd64.exe.zip` archive.
+. If you downloaded the archive, unzip it with a ZIP program.
```
In multiple procedures, you have this conditional phrase in step 2 `If you <did a specific action>, <do another specific action as a result>.` This formula is tripping me up because as a user, I'm thinking, "do I need to follow the other steps if I didn't download the archive? If I didn't download the archive, am I done after step one?" 

Like in my earlier comments, I think this whole procedure wants users to download the archive file, so I would simply tell them to do so in step one.

gabriel-rh replied:
I have added separate sub-steps for binary vs tarball.

As above, I agree that we should only document archives in future, but don't feel I should make that decision as I may not be fully aware of all the pros/cons.

gabriel-rh commented:
> @gabriel-rh, this looks good to me. But I have a few questions (not requesting changes).
> 
> * Since it's d.o.c, should installation instructions for RHEL be at the top instead of bottom?
>   
>   * If this makes sense, should "Installing odo on Linux" be reworded to something like "Installing odo on other Linux distributions"?
> * The table for "Installing odo on Linux" section looks neat. Why not have that for Windows & macOS as well? My reasoning for this is that, as a user, searching for the correct binary on the [content gateway](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/) is not a very pleasant experience.
> * Just like you have used the link for "OpenShift VS Code extension" in "Installing odo on VS Code" section, could you use that for the [content gateway link](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/) as well? Unless, of course, there's a reason to use the full URL

@dharmit I have added the tables to all sections, and replaced the full url with user-friendly text.

The structure matches that used for oc, and given the discussion above, I will leave it as is.

jc-berger reviewed (approved):
/lgtm

Preeticp commented:
@gabriel-rh looks great, loved the tables. Thank you!

Preeticp commented:
/cherrypick enterprise-4.6

Preeticp commented:
/cherrypick enterprise-4.7

Preeticp commented:
/cherrypick enterprise-4.8

Preeticp commented:
/cherrypick enterprise-4.9

Preeticp commented:
/cherrypick enterprise-4.10

Preeticp commented:
@gabriel-rh the cherrypicking of the PR worked for 4.9 and 4.10, but we will need to d a manual CP for the 4.6, 4.7 and 4.8 versions.
